File: cpp/matrix.cpp
// -------------------------------------------------------------------------------------------------- //
// - Program:     Matrix                                                                            - //
// - Paraneters:  ---                                                                               - //
// - Description: Determine the line numbers, the arithmetic average of elements that are less than - //
// -              the specified value.                                                              - //
// -------------------------------------------------------------------------------------------------- //

#include "utils.cpp"

class Matrix
{
    private:
        int count_rows;
        int count_cols;
        int count_all;

        int** table;
        int* sums;

        void initialize()
        {
            cout << "Matrix (size = " << count_rows << "x" << count_cols << ", total = " << count_all << "):" << endl;

            srand(time(NULL));
            table = new int*[count_rows];
            sums = new int[count_rows];
            for (int i = 0; i < count_rows; i++)
            {
                table[i] = new int[count_cols];
                sums[i] = 0;
                for (int j = 0; j < count_cols; j++)
                {
                    table[i][j] = randomize(0, 100);
                    sums[i] += table[i][j];
                    cout << '\t' << table[i][j];
                }
                sums[i] = (int) (sums[i] / count_cols);
                println();
            }
            println();
        }

    public:
        Matrix(int count_rows, int count_cols)
        {
            this->count_rows = count_rows;
            this->count_cols = count_cols;
            this->count_all = count_rows * count_cols;
            initialize();
        }

        ~Matrix()
        {
            for (int i = 0; i < count_rows; i++)
            {
                delete table[i];
                table[i] = nullptr;
            }
            delete[] table;
            table = nullptr;
            delete[] sums;
            sums = nullptr;
        }

        void averages()
        {
            cout << "Row average values:" << endl;
            for (int i = 0; i < count_rows; i++)
            {
                cout << '\t' << i + 1 << " -> " << sums[i] << endl;
            }
            println();
        }

        void result(int avg_number)
        {
            bool flag = false;
            cout << "Rows comply the condition:" << endl;
            for (int i = 0; i < count_rows; i++)
            {
                if (avg_number > sums[i])
                {
                    cout << '\t' << i + 1 << " -> " << sums[i] << endl;
                    flag = true;
                }
            }
            if (!flag)
            {
                cout << '\t' << "No such rows" << endl;
            }
        }
};

int main()
{
    int count_rows;
    cout << "Count of matrix rows: ";
    cin >> count_rows;
    rangeCheck(1, 20, count_rows);

    int count_cols;
    cout << "Count of matrix cols: ";
    cin >> count_cols;
    rangeCheck(1, 20, count_cols);

    println();

    Matrix *matrix = new Matrix(count_rows, count_cols);

    matrix->averages();

    int avg_number;
    cout << "Number for average values comparison: ";
    cin >> avg_number;
    rangeCheck(0, 101, avg_number);

    println();

    matrix->result(avg_number);

    delete matrix;
    matrix = nullptr;

    return 0;
}

File: csharp/matrix.cs
// -------------------------------------------------------------------------------------------------- //
// - Program:     Matrix                                                                            - //
// - Parameters:  ---                                                                               - //
// - Description: Determine the line numbers, the arithmetic average of elements that are less than - //
// -              the specified value.                                                              - //
// -------------------------------------------------------------------------------------------------- //

using System;

public class Matrix
{
    public int CountRows { get; private set; }
    public int CountCols { get; private set; }
    public int CountAll => CountRows * CountCols;
    public int[][] Table;
    public int[] Sums;

    public Matrix(int countRows, int countCols)
    {
        CountRows = countRows;
        CountCols = countCols;
        Initialize();
    }

    private void Initialize()
    {
        Console.WriteLine("Matrix (size = " + CountRows + "x" + CountCols + ", total = " + CountAll + "):");

        var rand = new Random();
        Table = new int[CountRows][];
        Sums = new int[CountRows];
        for (var i = 0; i < CountRows; i++)
        {
            Table[i] = new int[CountCols];
            Sums[i] = 0;
            for (var j = 0; j < CountCols; j++)
            {
                Table[i][j] = rand.Next(0, 100);
                Sums[i] += Table[i][j];
                Console.Write("\t" + Table[i][j]);
            }
            Sums[i] = (int) (Sums[i] / CountCols);
            Console.WriteLine();
        }
        Console.WriteLine();
    }

    public void Averages()
    {
        Console.WriteLine("Row average values:");
        for (var i = 0; i < CountRows; i++)
        {
            Console.WriteLine("\t" + (i + 1) + " -> " + Sums[i]);
        }
        Console.WriteLine();
    }

    public void Result(int avgNumber)
    {
        var flag = false;
        Console.WriteLine("Rows comply the condition (x < " + avgNumber + "):");
        for (var i = 0; i < CountRows; i++)
        {
            if (avgNumber > Sums[i])
            {
                Console.WriteLine("\t" + (i + 1) + " -> " + Sums[i]);
                flag = true;
            }
        }
        if (!flag)
        {
            Console.WriteLine("\t" + "No such rows");
        }
    }
}

static class Program
{
    static void Main(string[] args)
    {
        Console.Write("Count of matrix rows: ");
        int.TryParse(Console.ReadLine(), out var countRows);
        Utils.RangeCheck(1, 1000, countRows);

        Console.Write("Count of matrix cols: ");
        int.TryParse(Console.ReadLine(), out var countCols);
        Utils.RangeCheck(1, 1000, countCols);

        Console.WriteLine();

        var matrix = new Matrix(countRows, countCols);

        matrix.Averages();

        Console.Write("Number for average values comparison: ");
        int.TryParse(Console.ReadLine(), out var avgNumber);
        Utils.RangeCheck(0, 101, avgNumber);

        Console.WriteLine();

        matrix.Result(avgNumber);
    }
}